### PR TITLE
CSP-711 apprentice leaving the network

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.ApprenticeAan.Web/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    {
+        public const string MemberId = nameof(MemberId);
+        public const string MemberStatus = nameof(MemberStatus);
+    }
+
+    public static class MemberStatus
+    {
+        public static string Live = "Live";
+        public static string Withdrawn = "Withdrawn";
+    }
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Constants.cs:15]: So we need these constants, is there no enum?

File: src/SFA.DAS.ApprenticeAan.Web/Infrastructure/RouteNames.cs [hunk 2]
```diff
@@ -5,7 +5,6 @@ namespace SFA.DAS.ApprenticeAan.Web.Infrastructure;
 [ExcludeFromCodeCoverage]
 public static class RouteNames
 {
-    public const string Home = nameof(Home);
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Infrastructure/RouteNames.cs:8]: Why is this removed?

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+﻿using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using SFA.DAS.Aan.SharedUi.Infrastructure;
+using SFA.DAS.Aan.SharedUi.Models.LeaveTheNetwork;
+
+namespace SFA.DAS.ApprenticeAan.Web.Controllers;
+
+[Authorize]
+[Route("leaving-the-network-complete", Name = SharedRouteNames.LeaveTheNetworkComplete)]
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs:9]: leaving-network-confirmation

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+public class LeavingTheNetworkConfirmationController : Controller
+{
+    public const string LeaveTheNetworkConfirmedViewPath = "~/Views/LeaveTheNetwork/LeavingConfirmed.cshtml";
+
+    [HttpGet]
+    public IActionResult LeavingNetworkComplete()
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs:15]: To be consistent, this method name can be Index and then view name can match that up. So it can work conventionally. As long as the method name and view name are same, ASP will find it.

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/LeaveTheNetworkController.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            ReasonsForLeaving = reasonsTicked
+        };
+        _sessionService.Set(sessionModel);
+
+        return RedirectToRoute(SharedRouteNames.LeaveTheNetworkConfirmation);
+    }
+
+    [HttpGet]
+    [Route("leave-the-network-confirmation", Name = SharedRouteNames.LeaveTheNetworkConfirmation)]
+    public IActionResult AreYouSure()
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/LeaveTheNetworkController.cs:75]: AreYouSureGet

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/LeaveTheNetworkController.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+namespace SFA.DAS.ApprenticeAan.Web.Controllers;
+
+[Authorize]
+public class LeaveTheNetworkController : Controller
+{
+    private readonly IOuterApiClient _apiClient;
+    private readonly ISessionService _sessionService;
+
+    public const string LeaveTheNetworkViewPath = "~/Views/LeaveTheNetwork/Index.cshtml";
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/LeaveTheNetworkController.cs:19]: This will work conventionally, no need for explicit path.

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Extensions/SessionServiceExtensionsTests.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        var sessionServiceMock = new Mock<ISessionService>();
+        var actualMemberId = sessionServiceMock.Object.GetMemberId();
+        actualMemberId.Should().Be(Guid.Empty);
+    }
+
+    // [TestCase(true, true, true)]
+    [TestCase(false, true, false)]
+    //[TestCase(true, false, false)]
+    //[TestCase(false, false, false)]
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Extensions/SessionServiceExtensionsTests.cs:30]: Very last thing I did, forgot to uncomment.  Will do.

File: src/SFA.DAS.Aan.SharedUi/Models/LeaveTheNetwork/SubmitLeaveTheNetworkViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+﻿namespace SFA.DAS.Aan.SharedUi.Models.LeaveTheNetwork;
+public class SubmitLeaveTheNetworkViewModel
+{
+    public string LeavingReasonsTitle { get; set; } = null!;
+    public string LeavingExperienceTitle { get; set; } = null!;
+    public string LeavingBenefitsTitle { get; set; } = null!;
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.Aan.SharedUi/Models/LeaveTheNetwork/SubmitLeaveTheNetworkViewModel.cs:6]: no, I'll move to the get viewmodel

File: src/SFA.DAS.Aan.SharedUi/Models/LeaveTheNetwork/SubmitLeaveTheNetworkViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+    public List<LeavingReasonModel>? LeavingReasons { get; set; } = null!;
+    public List<LeavingReasonModel>? LeavingExperience { get; set; } = null!;
+    public List<LeavingReasonModel>? LeavingBenefits { get; set; } = null!;
+
+    public int LeavingExperienceRating { get; set; } = 0;
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.Aan.SharedUi/Models/LeaveTheNetwork/SubmitLeaveTheNetworkViewModel.cs:12]: ok

File: src/SFA.DAS.ApprenticeAan.Domain/Interfaces/ISessionService.cs [hunk 10]
```diff
@@ -2,8 +2,10 @@
 
 public interface ISessionService
 {
+    void Set(string key, string value);
     void Set<T>(T model);
     T Get<T>();
+    string? Get(string key);
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Domain/Interfaces/ISessionService.cs:8]: I'll put it in anyway

File: src/SFA.DAS.ApprenticeAan.Web/Infrastructure/RouteNames.cs (hunk 2 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Infrastructure/RouteNames.cs:8]: I moved it to SharedRouteNames

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs (hunk 3 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs:9]: ok

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/LeaveTheNetworkController.cs (hunk 5 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/LeaveTheNetworkController.cs:75]: ok

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/LeaveTheNetworkController.cs (hunk 6 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/LeaveTheNetworkController.cs:19]: ok

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs (hunk 4 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs:15]: ok

File: src/SFA.DAS.ApprenticeAan.Web/Filters/RequiresExistingMemberAttribute.cs [hunk 11]
```diff
@@ -28,12 +38,36 @@ private bool BypassCheck(ControllerActionDescriptor controllerActionDescriptor)
...
+        if (BypassCheck(controllerActionDescriptor)) return true;
+
+        var isMember = memberId != null;
 
         var isRequestingOnboardingPage = IsRequestForOnboardingAction(controllerActionDescriptor);
 
-        return (isMember && !isRequestingOnboardingPage) || (!isMember && isRequestingOnboardingPage);
+        return (isMember && isLive && !isRequestingOnboardingPage)
+               || (!isMember && isRequestingOnboardingPage)
+               || (isMember && !isLive && isRequestingOnboardingPage);
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Filters/RequiresExistingMemberAttribute.cs:69]: This actually simplifies out in the rejoin story, it's only grown because I needed to deal with what to do with a non-live member, when there is currently no rejoin page.  I will add coverage, but I'll do it in the follow up rejoin story.

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs (hunk 3 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/LeavingTheNetworkConfirmationController.cs:9]: ok

File: src/SFA.DAS.ApprenticeAan.Web/Constants.cs (hunk 1 above)
Comment (MarkFCain) [reply src/SFA.DAS.ApprenticeAan.Web/Constants.cs:15]: changed to an enum

File: src/SFA.DAS.Aan.SharedUi/Views/LeaveTheNetwork/Index.cshtml [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                            }
+                        </div>
+                        <p />
+                        <div id="LeavingExperienceRating" class="govuk-form-group">
+                            <h3 class="govuk-heading-m govuk-!-margin-bottom-4">@Model.LeavingExperienceTitle</h3>
+                            <div class="govuk-radios" data-module="govuk-radios">
+                                @{
+                                    for (var i = 0; i < Model.LeavingExperience!.Count; i++)
+                                    {
+                                        var forVariable = Model.LeavingExperience[i].Description;
```
Comment (DevenKShah) [inline src/SFA.DAS.Aan.SharedUi/Views/LeaveTheNetwork/Index.cshtml:51]: Wouldn't item id be better in this case  then the whole description as this is going to be id of the element

**Review by DevenKShah — APPROVED** [review]

File: src/SFA.DAS.Aan.SharedUi/Views/LeaveTheNetwork/Index.cshtml (hunk 12 above)
Comment (MarkFCain) [reply src/SFA.DAS.Aan.SharedUi/Views/LeaveTheNetwork/Index.cshtml:51]: I played around with this when I first did it, and for some reason it wouldn't work right for radiobuttons.  I will leave it for now, but revisit it in the next story
